chifra state crashes with null pointer

I was trying to retrieve the current value of the UnchainedIndex manifest variable and got a core dump.

This command

```
chifra state unchainedindex.eth --call '0x7087e4bd("0xf503017d7baf7fbc0fff7492b751025c6a78179b","mainnet")'
```

returns

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x1006f05ad]

goroutine 1 [running]:
github.com/TrueBlocks/trueblocks-core/src/apps/chifra/pkg/parser.(*ContractCallArgument).AbiType(0x0?, 0x0?)
	/Users/jrush/Development/trueblocks-core/src/apps/chifra/pkg/parser/contract_call.go:126 +0x20d
github.com/TrueBlocks/trueblocks-core/src/apps/chifra/pkg/call.convertArguments({0xc0005c20f0, 0x2, 0xa?}, 0xc0005c20f0?)
	/Users/jrush/Development/trueblocks-core/src/apps/chifra/pkg/call/call.go:119 +0x1ca
github.com/TrueBlocks/trueblocks-core/src/apps/chifra/pkg/call.NewContractCall({0xc00021f5c0, 0x7}, {{0xc, 0x31, 0x6b, 0x70, 0x42, 0xb4, 0x19, 0xd0, ...}}, ...)
	/Users/jrush/Development/trueblocks-core/src/apps/chifra/pkg/call/call.go:90 +0x399
github.com/TrueBlocks/trueblocks-core/src/apps/chifra/internal/state.(*StateOptions).HandleCall(0x1012d5920)
	/Users/jrush/Development/trueblocks-core/src/apps/chifra/internal/state/handle_call.go:24 +0x21a
github.com/TrueBlocks/trueblocks-core/src/apps/chifra/internal/state.(*StateOptions).StateInternal(0x1012d5920)
	/Users/jrush/Development/trueblocks-core/src/apps/chifra/internal/state/output.go:60 +0x36
github.com/TrueBlocks/trueblocks-core/src/apps/chifra/internal/state.RunState(0x10096a660?, {0xc00032c210?, 0x100a8c81f?, 0x8?})
	/Users/jrush/Development/trueblocks-core/src/apps/chifra/internal/state/output.go:30 +0x71
github.com/TrueBlocks/trueblocks-core/src/apps/chifra/pkg/file.RunWithFileSupport.func1(0x1012c09e0?, {0xc00032c210?, 0x1, 0x3})
	/Users/jrush/Development/trueblocks-core/src/apps/chifra/pkg/file/commands.go:110 +0x18c
github.com/spf13/cobra.(*Command).execute(0x1012c09e0, {0xc00032c1e0, 0x3, 0x3})
	/Users/jrush/go/pkg/mod/github.com/spf13/cobra@v1.2.1/command.go:856 +0x67c
github.com/spf13/cobra.(*Command).ExecuteC(0x1012c0260)
	/Users/jrush/go/pkg/mod/github.com/spf13/cobra@v1.2.1/command.go:974 +0x3bd
github.com/spf13/cobra.(*Command).Execute(...)
	/Users/jrush/go/pkg/mod/github.com/spf13/cobra@v1.2.1/command.go:902
github.com/TrueBlocks/trueblocks-core/src/apps/chifra/cmd.Execute()
	/Users/jrush/Development/trueblocks-core/src/apps/chifra/cmd/root.go:25 +0x25
main.main()
	/Users/jrush/Development/trueblocks-core/src/apps/chifra/main.go:33 +0xae
```

--- a/src/apps/chifra/pkg/abi/find.go
+++ b/src/apps/chifra/pkg/abi/find.go
@@ -19,16 +19,61 @@ const (
 func FindAbiFunction(mode FindMode, identifier string, arguments []*parser.ContractCallArgument, abis AbiInterfaceMap) (fn *types.SimpleFunction, suggestions []types.SimpleFunction, err error) {
 	for _, function := range abis {
 		function := function
-		// TODO: is this too naive?
+		// first we check if names match (or selectors, depending on mode)
 		if mode == FindByName && function.Name != identifier {
 			continue
 		}
 		if mode == FindBySelector && function.Encoding != strings.ToLower(identifier) {
 			continue
 		}
-		if arguments != nil && len(function.Inputs) != len(arguments) {
-			suggestions = append(suggestions, *function)
-			continue
+		// now we will compare arguments
+		if arguments != nil {
+			// we start with argument count
+			if len(function.Inputs) != len(arguments) {
+				suggestions = append(suggestions, *function)
+				continue
+			}
+			// now we check if the argument types are compatible
+			compatible := false
+			for index, input := range function.Inputs {
+				argument := arguments[index]
+				if input.InternalType == "string" && argument.String == nil {
+					break
+				}
+				if input.InternalType == "bool" && argument.Boolean == nil {
+					break
+				}
+				// address in parsed into argument.Hex.Address
+				if input.InternalType == "address" {
+					if argument.Hex == nil {
+						break
+					}
+					if argument.Hex.Address == nil {
+						break
+					}
+				}
+				// bytes32, array and tuples of any kind are only support via
+				// hashes. Hashes are parsed into argument.Hex.String
+				if input.InternalType == "bytes32" ||
+					strings.Contains(input.InternalType, "[") ||
+					strings.Contains(input.InternalType, "(") {
+					if argument.Hex == nil {
+						break
+					}
+					if argument.Hex.String == nil {
+						break
+					}
+				}
+				// we use strings.Contains here, because all integers are parsed into argument.Number
+				if strings.Contains(input.InternalType, "int") && argument.Number == nil {
+					break
+				}
+				compatible = true
+			}
+			if !compatible {
+				suggestions = append(suggestions, *function)
+				continue
+			}
 		}
 
 		return function, nil, nil

--- a/src/apps/chifra/pkg/abi/find_test.go
+++ b/src/apps/chifra/pkg/abi/find_test.go
@@ -4,6 +4,7 @@ import (
 	"strings"
 	"testing"
 
+	"github.com/TrueBlocks/trueblocks-core/src/apps/chifra/pkg/base"
 	"github.com/TrueBlocks/trueblocks-core/src/apps/chifra/pkg/parser"
 	"github.com/TrueBlocks/trueblocks-core/src/apps/chifra/pkg/types"
 	"github.com/TrueBlocks/trueblocks-core/src/apps/chifra/pkg/utils"
@@ -95,16 +96,20 @@ func init() {
 		panic(err)
 	}
 
-	abis = make(AbiInterfaceMap, len(testAbi.Methods))
-	for _, method := range testAbi.Methods {
-		method := method
-		encoding := "0x" + strings.ToLower(common.Bytes2Hex(method.ID))
-		abis[encoding] = types.FunctionFromAbiMethod(&method)
-	}
+	helperLoadAbisFromJson(&testAbi, &abis)
 
 	packTestAbi, err = abi.JSON(strings.NewReader(packTestAbiSource))
 	if err != nil {
 		panic(err)
+	}
+}
+
+func helperLoadAbisFromJson(parsedAbi *abi.ABI, destination *map[string]*types.SimpleFunction) {
+	*destination = make(AbiInterfaceMap, len(parsedAbi.Methods))
+	for _, method := range parsedAbi.Methods {
+		method := method
+		encoding := "0x" + strings.ToLower(common.Bytes2Hex(method.ID))
+		(*destination)[encoding] = types.FunctionFromAbiMethod(&method)
 	}
 }
 
@@ -247,5 +252,53 @@ func Test_findAbiFunctionBySelector(t *testing.T) {
 
 	if len(hints) != 0 {
 		t.Fatal("expected no hints")
+	}
+}
+
+func Test_findAbiFunctionMisleading(t *testing.T) {
+	// This test makes sure we choose the right function
+	// when given two very similar ones. In this case, it's
+	// transfer(bytes32, address) and transfer(address, address).
+	parsedAbi, err := abi.JSON(strings.NewReader(`
+		[
+			{ "type" : "function", "name" : "transfer", "inputs" : [ { "name" : "from", "type" : "bytes32" }, { "name" : "to", "type" : "address" } ] },
+			{ "type" : "function", "name" : "transfer", "inputs" : [ { "name" : "from", "type" : "address" }, { "name" : "to", "type" : "address" } ] }
+		]
+	`))
+	if err != nil {
+		t.Fatal(err)
+	}
+	abis := make(map[string]*types.SimpleFunction)
+	helperLoadAbisFromJson(&parsedAbi, &abis)
+
+	call := &parser.FunctionContractCall{
+		Name: "transfer",
+		Arguments: []*parser.ContractCallArgument{
+			{
+				Hex: &parser.ContractCallHex{
+					Address: utils.PointerOf(base.HexToAddress("0x6B175474E89094C44Da98b954EedeAC495271d0F")),
+				},
+			},
+			{
+				Hex: &parser.ContractCallHex{
+					Address: utils.PointerOf(base.HexToAddress("0x6B175474E89094C44Da98b954EedeAC495271d0F")),
+				},
+			},
+		},
+	}
+
+	result, hints, err := FindAbiFunction(FindByName, call.Name, call.Arguments, abis)
+	if err != nil {
+		t.Fatal(err)
+	}
+	wanted := parsedAbi.Methods["transfer0"]
+	if result == nil {
+		t.Fatal("did not find anything")
+	}
+	if result.Signature != wanted.Sig {
+		t.Fatal("found wrong function:", result.Signature)
+	}
+	if len(hints) > 0 {
+		t.Fatal("expected no hints:", hints)
 	}
 }

--- a/src/apps/chifra/pkg/parser/contract_call.go
+++ b/src/apps/chifra/pkg/parser/contract_call.go
@@ -4,6 +4,7 @@ import (
 	"errors"
 	"fmt"
 	"math/big"
+	"reflect"
 	"strconv"
 
 	"github.com/TrueBlocks/trueblocks-core/src/apps/chifra/pkg/base"
@@ -66,6 +67,8 @@ type FunctionContractCall struct {
 // ContractCallArgument represents input to the smart contract method call, e.g.
 // `true` in `setSomething(true)`
 type ContractCallArgument struct {
+	Tokens []lexer.Token
+
 	String  *string             `parser:"@String"`
 	Number  *ContractCallNumber `parser:"| @Decimal"`
 	Boolean *Boolean            `parser:"| @('true'|'false')"`
@@ -94,7 +97,10 @@ func (a *ContractCallArgument) Interface() any {
 
 func (a *ContractCallArgument) AbiType(abiType *abi.Type) (any, error) {
 	if abiType.T == abi.FixedBytesTy {
-		// We only support fixed bytes as hex strings
+		// We only support fixed bytes as hashes
+		if a.Hex == nil {
+			return nil, newWrongTypeError("hash", a.Tokens[0], a.Interface())
+		}
 		hex := *a.Hex.String
 		if len(hex) == 0 {
 			return nil, errors.New("no value for fixed-size bytes argument")
@@ -108,6 +114,9 @@ func (a *ContractCallArgument) AbiType(abiType *abi.Type) (any, error) {
 	}
 
 	if abiType.T == abi.IntTy {
+		if a.Number == nil {
+			return nil, newWrongTypeError(abiType.String(), a.Tokens[0], a.Interface())
+		}
 		// We have to convert int64 to a correct int type, otherwise go-ethereum will
 		// return an error. It's not needed for uints, because they handle them differently.
 		if a.Number.Big != nil {
@@ -123,13 +132,23 @@ func (a *ContractCallArgument) AbiType(abiType *abi.Type) (any, error) {
 
 	if abiType.T == abi.AddressTy {
 		// We need go-ethereum's Address type, not ours
+		if a.Hex == nil {
+			return nil, newWrongTypeError(abiType.String(), a.Tokens[0], a.Interface())
+		}
 		address := a.Hex.Address
 		if address == nil {
-			return nil, errors.New("expected address")
+			return nil, errors.New("expected address, but got hash instead (check length)")
 		}
 
 		addressHex := common.HexToAddress(address.Hex())
 		return addressHex, nil
+	}
+
+	// Below checks are for nice errors only
+	if (abiType.T == abi.UintTy && a.Number == nil) ||
+		(abiType.T == abi.StringTy && a.String == nil) ||
+		(abiType.T == abi.BoolTy && a.Boolean == nil) {
+		return nil, newWrongTypeError(abiType.String(), a.Tokens[0], a.Interface())
 	}
 
 	return a.Interface(), nil
@@ -254,6 +273,17 @@ func (s *Selector) Capture(values []string) error {
 
 	s.Value = literal
 	return nil
+}
+
+func newWrongTypeError(expectedType string, token lexer.Token, value any) error {
+	t := reflect.TypeOf(value)
+	typeName := t.String()
+	kind := t.Kind()
+	// kinds between this range are all (u)int, called "integer" in Solidity
+	if kind > 1 && kind < 12 {
+		typeName = "integer"
+	}
+	return fmt.Errorf("expected %s, but got %s \"%s\"", expectedType, typeName, token)
 }
 
 // Build parser

--- a/src/apps/chifra/pkg/parser/contract_call_test.go
+++ b/src/apps/chifra/pkg/parser/contract_call_test.go
@@ -8,6 +8,7 @@ import (
 
 	"github.com/TrueBlocks/trueblocks-core/src/apps/chifra/pkg/base"
 	"github.com/TrueBlocks/trueblocks-core/src/apps/chifra/pkg/utils"
+	"github.com/alecthomas/participle/v2/lexer"
 	"github.com/ethereum/go-ethereum/accounts/abi"
 	"github.com/ethereum/go-ethereum/common"
 )
@@ -197,7 +198,7 @@ const packTestAbiSource = `
 	{ "type" : "function", "name" : "bytes32", "inputs" : [ { "name" : "inputs", "type" : "bytes32" } ] }
 ]`
 
-func TestArgument_ToAbiType(t *testing.T) {
+func TestArgument_AbiType(t *testing.T) {
 	testAbi, err := abi.JSON(strings.NewReader(packTestAbiSource))
 	if err != nil {
 		panic(err)
@@ -302,5 +303,45 @@ func TestArgument_ToAbiType(t *testing.T) {
 				t.Errorf("Argument.ToAbiType() = %v, want %v", got, tt.want)
 			}
 		})
+	}
+}
+
+func TestArgument_AbiType_Errors(t *testing.T) {
+	testAbi, err := abi.JSON(strings.NewReader(packTestAbiSource))
+	if err != nil {
+		panic(err)
+	}
+	// the second argument is string instead of address
+	parsed, err := ParseContractCall(`transfer(0x6982508145454ce325ddbe47a25d4ec3d23119a1, "0x6982508145454ce325ddbe47a25d4ec3d23119a1")`)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if _, err := parsed.FunctionNameCall.Arguments[0].AbiType(&testAbi.Methods["address"].Inputs[0].Type); err != nil {
+		t.Fatal(err)
+	}
+	_, err = parsed.FunctionNameCall.Arguments[1].AbiType(&testAbi.Methods["address"].Inputs[0].Type)
+	expectedError := newWrongTypeError("address", lexer.Token{Value: `0x6982508145454ce325ddbe47a25d4ec3d23119a1`}, "0x6982508145454ce325ddbe47a25d4ec3d23119a1")
+	if err.Error() != expectedError.Error() {
+		t.Fatal("got wrong error:", err, "expected:", expectedError)
+	}
+
+	parsed, err = ParseContractCall(`someBool(111)`)
+	if err != nil {
+		t.Fatal(err)
+	}
+	_, err = parsed.FunctionNameCall.Arguments[0].AbiType(&testAbi.Methods["bool"].Inputs[0].Type)
+	expectedError = newWrongTypeError("bool", lexer.Token{Value: `111`}, 111)
+	if err.Error() != expectedError.Error() {
+		t.Fatal("got wrong error:", err, "expected:", expectedError)
+	}
+
+	parsed, err = ParseContractCall(`someBytes32("hello")`)
+	if err != nil {
+		t.Fatal(err)
+	}
+	_, err = parsed.FunctionNameCall.Arguments[0].AbiType(&testAbi.Methods["bytes32"].Inputs[0].Type)
+	expectedError = newWrongTypeError("hash", lexer.Token{Value: `hello`}, "hello")
+	if err.Error() != expectedError.Error() {
+		t.Fatal("got wrong error:", err, "expected:", expectedError)
 	}
 }

